### PR TITLE
converting shell file to python with latest modification

### PR DESCRIPTION
**Pull request by Naresh-ibm**
re-writting the script from shell to python. handled the
dependency module unload without giving them in config file.
for FC adapter multipath dependency is handled properly.
code can be run for a single module or for all the driver
modules available in the OS

Signed-off-by: Naresh Bannoth <nbannoth@in.ibm.com>

### CODE REVIEW

**Comment by Naresh-ibm** [comment]
[module_unload_load_Sailfish_rhel8.1_zz1_job_log.txt](https://github.com/avocado-framework-tests/avocado-misc-tests/files/3486440/module_unload_load_Sailfish_rhel8.1_zz1_job_log.txt)

File: io/driver/module_unload_load.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+from avocado.utils import process
+from avocado.utils import linux_modules
+from avocado.utils import pci
+from avocado import Test
+
+
+class ModuleLoadUnload(Test):
+
+    """
+    PCI devices can be bound and unbound to drivers.
```
Comment (narasimhan-v) [inline io/driver/module_unload_load.py]: `Drivers can be loaded and unloaded`

File: io/driver/module_unload_load.py [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    :param iteration: Number of time to unload and load the module
+    """
+    def setUp(self):
+        """
+        get parameters.
+        """
+        self.module = self.params.get('module', default=None)
+        self.iteration = self.params.get('iteration', default=1)
+        self.only_io = self.params.get('only_io', default=None)
+        self.fc = self.params.get('fc', default=None)
```
Comment (narasimhan-v) [inline io/driver/module_unload_load.py]: Consider handling this in a more generic way.

File: io/driver/module_unload_load.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+    def built_in_module(self, module):
+        """
+        checking whether the given module is built_in module or not
+        """
+        uname = process.getoutput("uname -r")
```
Comment (narasimhan-v) [inline io/driver/module_unload_load.py]: Consider using utils

File: io/driver/module_unload_load.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+        cmd = "cat /lib/modules/%s/modules.builtin" % uname
```
Comment (narasimhan-v) [inline io/driver/module_unload_load.py]: Consider using utils

File: io/driver/module_unload_load.py [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+            return False
+
+    def module_load_unload(self, module):
+        """
+        Unloading and loading the given module
+        """
+        if self.built_in_module(module) is True and self.only_io is True:
+            self.fail("%s is built_in Module can't be unloaded" % module)
+        else:
+            self.log.info("%s is built_in Module can't unload" % module)
```
Comment (narasimhan-v) [inline io/driver/module_unload_load.py]: this block can be handled in a way it is easier to understand.

File: io/driver/module_unload_load.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+        self.sub_modules.append(module)
+        for _ in range(0, self.iteration):
+            for mdl in self.sub_modules:
+                if self.fc is True:
+                    process.system("multipath -F", ignore_status=True)
+                    time.sleep(5)
+                self.log.info("unloading : %s " % mdl)
+                process.system("multipath -F", ignore_status=True)
+                time.sleep(5)
+                process.system("rmmod %s" % mdl)
```
Comment (narasimhan-v) [inline io/driver/module_unload_load.py]: Consider using utils

File: io/driver/module_unload_load.py [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+            time.sleep(5)
+        self.sub_modules = linux_modules.get_submodules(module)
+        self.sub_modules.append(module)
+        for _ in range(0, self.iteration):
+            for mdl in self.sub_modules:
+                if self.fc is True:
+                    process.system("multipath -F", ignore_status=True)
+                    time.sleep(5)
+                self.log.info("unloading : %s " % mdl)
+                process.system("multipath -F", ignore_status=True)
```
Comment (narasimhan-v) [inline io/driver/module_unload_load.py]: why twice ?

File: io/driver/module_unload_load.py [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+                time.sleep(5)
+                process.system("rmmod %s" % mdl)
+                time.sleep(5)
+            for mdl in reversed(self.sub_modules):
+                self.log.info("loading : %s " % mdl)
+                linux_modules.load_module(mdl)
+                time.sleep(3)
```
Comment (narasimhan-v) [inline io/driver/module_unload_load.py]: Can you make the sleep time uniform ?

File: io/driver/module_unload_load.py.data/module_unload_load.yaml [hunk 9]
```diff
@@ -0,0 +1,5 @@
+Parameters:
+    module: "qla2xxx"
```
Comment (narasimhan-v) [inline io/driver/module_unload_load.py.data/module_unload_load.yaml]: remove these default values.

**Comment by PraveenPenguin** [comment]
@Naresh-ibm @narasimhan-v just we make sure this should work with upstream avocado also

File: io/driver/module_unload_load.py (hunk 3 above)
Comment (Naresh-ibm) [reply io/driver/module_unload_load.py]: @narasimhan-v  can you please point me the .py file under utils?

Comment (narasimhan-v) [reply io/driver/module_unload_load.py]: @Naresh-ibm `grep uname avocado/utils/*.py`

File: io/driver/module_unload_load.py (hunk 4 above)
Comment (Naresh-ibm) [reply io/driver/module_unload_load.py]: genio.read(/lib/modules/%s/modules.builtin" % uname).split('/')[-1]

File: io/driver/module_unload_load.py (hunk 7 above)
Comment (Naresh-ibm) [reply io/driver/module_unload_load.py]: ifconfig file is not there, then ok, other wise add multipath,multipath in config file

File: io/driver/module_unload_load.py (hunk 8 above)
Comment (Naresh-ibm) [reply io/driver/module_unload_load.py]: use the static variable/time.wait/ add sleep time to utils
command will be 

unload_sleep_time = 15
load_sleep_time = 5
for unload    time.sleep unload_sleep_time
for  load      time.sleep load_sleep_time

File: io/driver/module_unload_load.py (hunk 6 above)
Comment (Naresh-ibm) [reply io/driver/module_unload_load.py]: open a issue for utils problem or use utils here.

File: io/driver/module_unload_load.py (hunk 5 above)
Comment (Naresh-ibm) [reply io/driver/module_unload_load.py]: prepare a list of module and handle accordinly.

**Comment by Naresh-ibm** [comment]
closing this in favour of #1289

**Pull request closed without merging**